Add support for arbritrary optimization

sstendahl opened the pull request:
This now works in principle, with a few complete examples in the accompanying notebook. 
The examples in the notebook might need some slight work.

TODO:

- [x] Fix docstrings 
- [x] Fix linting issues
- [x] Fix existing unit tests
- [x] Write unit tests
- [x] Add type hints to new methods
- [x] Probably want to implement support for setting the instrument in `optimise_parameters` already in this PR

jfkcooper commented on hogben/optimise.py:
```diff
@@ -12,6 +12,64 @@
+from hogben.visualise import scan_parameters
+
+
+def optimise_parameters(sample: BaseSample,
```
I think you have different formatting/indentation for these arguments to all other functions in the file

jfkcooper commented on hogben/optimise.py:
```diff
@@ -95,13 +153,13 @@ def _non_decreasing(x):
+            angle_splits: list,
+            total_time: float = 1000,
+            contrast_bounds: tuple = (-0.56, 6.36),
+            workers: int = -1,
```
for workers in the docstring, can you say what `-1` means

jfkcooper commented on hogben/optimise.py:
```diff
@@ -156,16 +214,80 @@ def _non_decreasing(x):
+            self,
+            angle_times,
+            inst_or_path='OFFSPEC',
+            workers=-1,
```
for workers in the docstring, can you say what `-1` means

jfkcooper commented on hogben/optimise.py:
```diff
@@ -156,16 +214,80 @@ def _non_decreasing(x):
+                          'OFFSPEC'
+
+        Returns:
+            float: negative of minimum eigenvalue using given conditions.
```
negative of minimum eigenvalue using given conditions. -->negative of minimum eigenvalue of the Fisher information matrix using the given conditions.

sstendahl replied:
In case you're not aware by the way, nowadays you can also add suggestions in GitHub, so that these can just be accepted by the person who submitted the PR with a press of a button like this:
``` 
```suggestion
            float: negative of minimum eigenvalue using given conditions.
\```
```
Without the backslash on the last ``` here. This results on something like this:

```suggestion
            float: negative of minimum eigenvalue using given conditions.
```

jfkcooper commented on hogben/models/samples.py:
```diff
@@ -55,29 +71,44 @@ def __vary_structure(structure, bound_size=0.2):
+                    sld = component.sld.real
+                    sld_bounds = (
+                        sld.value * (1 - bound_size),
+                        sld.value * (1 + bound_size),
```
I think we should check if parameters already have bounds defined before assigning them

sstendahl replied:
Probably makes sense. I'm not 100% sure why we need `vary_structure` so it kinda depends on what the use-case is. But I think in principle it makes sense not to override previous bounds.

I kept it there not to break old workflows and implementations/notebooks. I think at the moment I mainly use it in unit tests...

sstendahl replied:
So I just looked this up, and by default (when not assigned any bounds) it sets the bounds to positive and negative infinity. I changed this somewhat to only use these newly redefined bounds when the bound is (+/-) infinite for each respective bound.

```
                if sld.bounds.lb == -float("inf"):
                    sld.bounds.lb = sld.value * (1 - bound_size)
                if sld.bounds.ub == float("inf"):
                    sld.bounds.ub = sld.value * (1 + bound_size)

                sld_bounds = (
                    sld.bounds.lb,
                    sld.bounds.ub,
                )
```

I could do the `if` conditional on a single line for each one, but that become pretty ugly to read.

jfkcooper commented on hogben/tests/test_optimise.py:
```diff
@@ -135,3 +173,58 @@ def test_underlayers_func():
+    # Check whether the visualisation functions are called
+    sample.sld_profile.assert_called_once()
+    sample.reflectivity_profile.assert_called_once()
+    sample.reflectivity_profile.scan_parameters()
```
should this have an assert associated?

sstendahl replied:
Yes, it's a bit of a shallow test but it should just check whether scan_parameters is called. It's actually completely off, it should be `mock_scan_parameters.assert_called_once()`

jfkcooper commented on hogben/utils.py:
```diff
@@ -150,6 +150,27 @@ def __init__(self,
+                    inst_or_path='OFFSPEC'):
+        """
+        Get Fisher object using a sample.
+        Seperate constructor for magnetic simulation maybe? Probably depends
```
improve docstring and add comment as a TODO

sstendahl replied:
Oops, kinda missed that one. That was actually pretty old, the comment was actually from before I implemented magnetic layers locally. 

This can be done with this same constructor, so TODO not neccesary even here. But docstring needs to be fixed for sure, that was sloppy :(

jfkcooper reviewed (changes requested):
A few comments throughout, and I have some questions on the notebook, but maybe easier face to face if you have time at some point soon?

sstendahl commented on hogben/models/bilayers.py:
```diff
@@ -719,6 +731,16 @@ def __init__(self):
         super().__init__()
 
+    def get_structures(self) -> list:
+        """
```
Actually, it should. At the time, I was under the impression that `BilayerDPPC` was missing a native `structures` attribute. Hence creating it with `_using_conditions`. But this is actually already done on L748.

sstendahl commented on hogben/models/samples.py:
```diff
@@ -89,38 +120,39 @@ def angle_info(self, angle_times, contrasts=None):
         """
         # Return the Fisher information matrix calculated from simulated data.
-        model = refnx.reflect.ReflectModel(self.structure)
+        model = refnx.reflect.ReflectModel(self.get_structures()[0])
```
I see this is a half-arsed implementation. Basically, in the current main branch a `Sample` only has a single model. So the existing methods were written with only one model in mind. So this old method was just quickly adapted to keep working.

But this should be very fixable, as `Fisher` can be calculated from a list of models, qs and counts. So this method can be adapted to include all models.

sstendahl commented on hogben/models/samples.py:
```diff
@@ -89,38 +120,39 @@ def angle_info(self, angle_times, contrasts=None):
 
         Args:
             save_path (str): path to directory to save SLD profile to.
+            single (bool): whether to plot all profiles on a single plot or
```
Oops, yes I had that in an older version. It was not a great naming there anyway, but what the argument did was to show all SLD-profiles in a seperate plot if `single=False`. Might implement that later, but decided to not let that be part of this PR as the scope is wide enough as it is now anyway.

There were some issues with these multiple plots (mostly with overlapping text and sizing issues), hence just leaving the implementation out for now.

sstendahl commented on hogben/models/samples.py:
```diff
@@ -208,7 +249,7 @@ def nested_sampling(self,
         """
         # Simulate data for the sample.
-        model = refnx.reflect.ReflectModel(self.structure)
+        model = refnx.reflect.ReflectModel(self.get_structures()[0])
```
Same with `angle_info`, that the original code was written for a single structure. And I just used this bandaid solution to keep things working. Will take a look and see if this is generalizable to a list of structures, I think it should be.

sstendahl replied:
I think I've got a working version, but I keep getting a NaN error when it tries to calculate the log likelihoods. I'm getting this also in the main branch for a `Sample` object (it works for the Bilayers), so the behaviours is identical. Just need to get a closer look on what is needed for this to work properly.

sstendahl commented on hogben/models/samples.py:
```diff
@@ -55,29 +71,44 @@ def __vary_structure(structure, bound_size=0.2):
+        for structure in self.structures:
+            params = []
+            # The structure was defined in refnx.
+            if isinstance(structure, refnx.reflect.Structure):
```
Not really, but this was part in the original main code when `refl1d` was still supported. And I just kept it in.

sstendahl commented on hogben/models/base.py:
```diff
@@ -56,16 +55,61 @@ class BaseSample(VariableAngle):
+                if hasattr(p, attr) and getattr(p, attr):
+                    params.append(p)
+                    continue
+                if p._deps:
```
This is basically an adaption from the `varying_parameters` method in `refnx` ([link](https://github.com/refnx/refnx/blob/2fd237d10a8b5c065c186837e7e40821e1219be0/refnx/analysis/objective.py#L372)), but extended to allow for "custom" attributes as well instead of just the `vary` parameter:

```

    def varying_parameters(self):
        """
        Returns
        -------
        varying_parameters : refnx.analysis.Parameters
            The varying Parameter objects allowed to vary during the fit.

        """
        # create and return a Parameters object because it has the
        # __array__ method, which allows one to quickly get numerical values.
        lst = []
        for p in flatten(self.parameters):
            if p.vary:
                lst.append(p)
                continue
            if len(p._deps):
                lst.extend([_p for _p in p.dependencies() if _p.vary])
        # should already be totally flattened by this point
        return Parameters(f_unique(lst))
        ```
        
See also `p.dependencies`:

```

    def dependencies(self):
        dep_list = []
        for _dep in self._deps:
            if isinstance(_dep, Parameter):
                if _dep.constraint is not None:
                    dep_list.append(_dep.dependencies())
                else:
                    dep_list.append(_dep)
            if isinstance(_dep, (_UnaryOp, _BinaryOp)):
                dep_list.append(_dep.dependencies())
        return list(f_unique(flatten(dep_list)))
```

Where the second statement checks the nested dependencies, and adds these as well if the attributes match. I'll have to do a double check, but this could be for instance the iSLD and the real SLD. Will clarify a bit in the code itself.

sstendahl replied:
Did some investigation, and this indeed gets the parameters that are coupled to the model due to dependencies. So to clarify, imagine our sample is like: `substrate | layer1| layer2| air|`, and we've got the following parameters:
```
    gamma = Parameter(0.5, 'Gamma', (0, 1))
    period = Parameter(120, 'Period', (60, 190))
    layer1_thick = period*gamma
    layer2_thick = period*(1-gamma)
    gamma.vary = True
```

The `gamma` parameter will actually be some `BinaryOps` parameter that will not have the `vary` attribute. But this parameter will instead be part of the `parameter.dependencies` attribute. So what this second `if` basically does is adding the parameters that are not directly tied to an SLD-slab parameter, but instead tied to the model using such dependencies as the code-block in this comment above. I added a short comment here to clarify this somewhat.

sstendahl commented on hogben/models/base.py:
```diff
@@ -56,16 +55,61 @@ class BaseSample(VariableAngle):
+
+    def get_models(self) -> list:
+        """
+        Returns a list of all refnx `ReflectModel` models that are
```
I've changed the docstring. 

For the lipids, I've added a seperate implementation for `get_models`  in `BaseLipids`, where it takes advantage of the fact that these structures all have a separately defined scale and bkg. 

What I could do as well, is to modify the basic Sample (`samples.Sample`), such that it also has a list of `bkg`, `dq' and `scale`, corresponding to the total amount of structures. If a single value is given, it would then just create a list with the same values. And if a list is given it would use the list. If the total amount of elements do not correspond to the total amount of structures, an error should be raised. Will look into that before  resolving. That could deduplicate things a bit further.

sstendahl replied:
So I've changed it a bit more. Now as described, `dq`, `scale` and `bkg` can either be given as a single float/int. In that case the same value will be used for each structure. Or it can be given as a list, in case it will use the specified values in the same order as the structures.

For the magnetic cases (PR after this), I will likely keep it so that `scale`, `dq` and `bkg` will be the same for spin-up and spin-down. Meaning the same amount of parameters need to be given for polarised experiments as for unpolarised.

jfkcooper commented on notebooks/optimization.ipynb:
```diff
@@ -0,0 +1,388 @@
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 5
```
In this notebook, would it be possible to add how to use different angle_times for each contrast, and how to use a different background etc?
It is currently not clear where these should be fed in.

sstendahl replied:
I'll add that now. Apologies for being late, I was expecting an email notification with the requested changes but didn't get that. Will also add another unit tests to check the order, just added another few as well.

jfkcooper reviewed (changes requested):
I only added a single comment, but if you could also have a test to ensure that the scales/backgrounds are kept in the correct orders that would also be great.
Otherwise it is looking good!

sstendahl commented:
Apologies for the future creep here, but this should be ready now. I've changed the reflectivity graph after the optimization to show both the model reflectivity, as well as the simulated experiment with error bars.

![image](https://github.com/jfkcooper/HOGBEN/assets/68477016/2308bfe5-fa30-4466-9abf-300cce723bac)


I've also changed the label logic. If shows a label if there's more than one structure to display. By default it generates this automatically, using the solvent SLD as label if the solvents are different and otherwise it simply numbers the structures. Finally, I've added the ability to set the labels when initiating the sample as well. Which seems to work fine.  Added some unit tests for that as well.

jfkcooper reviewed (approved):
Looks good!
